### PR TITLE
Upgrade telegraf v1-23-0

### PR DESCRIPTION
**Pull request by ZhenyuTan-amz**
# Description of the issue
Switch to use the latest Telegraf agent, fix all the building errors. Bump Go version to 1.17

# Description of changes
Minor code changes to allow a success build

# License
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Tests

### CODE REVIEW

**Comment by pingleig** [comment]
I remember we had a fork on telegraf for lvm disk detection https://github.com/aws/telegraf/commits/master https://github.com/shirou/gopsutil/pull/957 and https://github.com/aws/amazon-cloudwatch-agent/pull/106

File: Makefile [hunk 1]
```diff
@@ -51,7 +51,7 @@ GOIMPORTS_OPT?= -w -local $(CW_AGENT_IMPORT_PATH)
 GOIMPORTS = $(TOOLS_BIN_DIR)/goimports
 SHFMT = $(TOOLS_BIN_DIR)/shfmt
 LINTER = $(TOOLS_BIN_DIR)/golangci-lint
-release: clean test build package-rpm package-deb package-win package-darwin
+release: clean test build package-rpm
```
Comment (adam-mateen) [inline Makefile:54]: This change to the makefile's "release" target seems unrelated.
Why was it necessary?

File: amazon-cloudwatch-container-insights/k8s-deployment-manifest-templates/deployment-mode/service/cwagent-prometheus/prometheus-eks.yaml [hunk 2]
```diff
@@ -13,6 +13,9 @@ data:
   # cwagent json config
   cwagentconfig.json: |
     {
+      "agent": {
+        "debug": true
+      },
```
Comment (adam-mateen) [inline amazon-cloudwatch-container-insights/k8s-deployment-manifest-templates/deployment-mode/service/cwagent-prometheus/prometheus-eks.yaml:18]: Setting `"debug" : true` in a template... Does not seem necessary.
In my opinion "debug" should not be a default.

File: amazon-cloudwatch-container-insights/k8s-deployment-manifest-templates/deployment-mode/service/cwagent-prometheus/prometheus-eks.yaml [hunk 3]
```diff
@@ -449,7 +448,7 @@ spec:
     spec:
       containers:
         - name: cloudwatch-agent
-          image: amazon/cloudwatch-agent:1.247346.0b249609
+          image: public.ecr.aws/a3i2u7g3/cw-agent-seth:upgrade-telegraf-v1-21-4
```
Comment (adam-mateen) [inline amazon-cloudwatch-container-insights/k8s-deployment-manifest-templates/deployment-mode/service/cwagent-prometheus/prometheus-eks.yaml:451]: Probably do not want "seth" specific image.

File: amazon-cloudwatch-container-insights/k8s-yaml-templates/cwagent-kubernetes-monitoring/cwagent-configmap.yaml [hunk 4]
```diff
@@ -5,10 +5,13 @@ data:
...
   cwagentconfig.json: |
     {
+      "agent": {
+        "debug": true
+      },
       "logs": {
         "metrics_collected": {
           "kubernetes": {
-            "cluster_name": "{{cluster_name}}",
+            "cluster_name": "seth-containerd-test-cluster",
```
Comment (adam-mateen) [inline amazon-cloudwatch-container-insights/k8s-yaml-templates/cwagent-kubernetes-monitoring/cwagent-configmap.yaml:14]: Was this a temporary change for testing that can now be removed?

File: cmd/start-amazon-cloudwatch-agent/start-amazon-cloudwatch-agent.go [hunk 5]
```diff
@@ -92,8 +93,25 @@ func main() {
...
+		log.Fatal(err)
+	}
+	defer func() {
+		if err = file.Close(); err != nil {
+			log.Fatal(err)
+		}
+	}()
+
+	b, err := ioutil.ReadAll(file)
+	log.Printf("I! toml config %v", string(b))
```
Comment (adam-mateen) [inline cmd/start-amazon-cloudwatch-agent/start-amazon-cloudwatch-agent.go:116]: nitpick on naming - I'd prefer `printFileContents()` or `logFileContents()`.

File: plugins/parsers/registry.go [hunk 6]
```diff
@@ -6,24 +6,25 @@ package parsers
...
 	}
-	return parser, nil
+	return parser
 }
 
 func init() {
-	parsers.RegisterParser("emf", NewEMFParser)
-	parsers.RegisterParser("aws_csm", NewCSMParser)
+	parsers.Add("emf", EMFParserCreator)
+	parsers.Add("aws_csm", CSMParserCreator)
```
Comment (adam-mateen) [inline plugins/parsers/registry.go:29]: CWA has a lot of code that names a constructor for struct "Foo" as `NewFoo()`.
I recommend following the convention and keeping the old names:
`NewEMFParser()` and `NewCSMParser()`.

File: amazon-cloudwatch-container-insights/cloudwatch-agent-dockerfile/source/Dockerfile [hunk 7]
```diff
@@ -6,12 +6,16 @@ FROM $GO_IMAGE as builder
 # NOTE: This arg will be populated by docker buildx
 # https://docs.docker.com/engine/reference/builder/#automatic-platform-args-in-the-global-scope
 ARG TARGETARCH
+ENV GOPROXY=direct
```
Comment (SaxyPandaBear) [inline amazon-cloudwatch-container-insights/cloudwatch-agent-dockerfile/source/Dockerfile:9]: Is the change to the go proxy necessary here? I'm assuming it is based on the `go mod download -x` further down in the Dockerfile. What are the consequences of this?

File: amazon-cloudwatch-container-insights/k8s-deployment-manifest-templates/deployment-mode/service/cwagent-prometheus/prometheus-eks.yaml (hunk 2 above)
Comment (SaxyPandaBear) [reply amazon-cloudwatch-container-insights/k8s-deployment-manifest-templates/deployment-mode/service/cwagent-prometheus/prometheus-eks.yaml:18]: Agreed. Also, the deployment YAMLs that need to be updated are in the Container Insights repo, not this one.

File: amazon-cloudwatch-container-insights/k8s-yaml-templates/cwagent-kubernetes-monitoring/cwagent-daemonset.yaml [hunk 8]
```diff
@@ -15,7 +15,8 @@ spec:
     spec:
       containers:
         - name: cloudwatch-agent
-          image: amazon/cloudwatch-agent:1.230621.0
+          image: public.ecr.aws/a3i2u7g3/cw-agent-seth:upgrade-telegraf-v1-21-4
```
Comment (SaxyPandaBear) [inline amazon-cloudwatch-container-insights/k8s-yaml-templates/cwagent-kubernetes-monitoring/cwagent-daemonset.yaml:18]: same sentiment as other comments. I understand that this is an image that works, but maybe what we need is to point to the `latest` image in ECR

File: amazon-cloudwatch-container-insights/k8s-yaml-templates/cwagent-kubernetes-monitoring/otel-config.yml [hunk 9]
```diff
@@ -0,0 +1,252 @@
+# create namespace
```
Comment (SaxyPandaBear) [inline amazon-cloudwatch-container-insights/k8s-yaml-templates/cwagent-kubernetes-monitoring/otel-config.yml:1]: should we add this to the Container Insights repo?

File: cmd/amazon-cloudwatch-agent/register_event_logger_notwindows.go [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+// SPDX-License-Identifier: MIT
+
+//go:build !windows
+// +build !windows
+
+package main
+
+func RegisterEventLogger() error {
+	// Unreachable code, do nothing
+	return nil
```
Comment (SaxyPandaBear) [inline cmd/amazon-cloudwatch-agent/register_event_logger_notwindows.go:11]: It would help me understand this change if there was a comment explaining why this doesn't do anything on non-Windows. Is this because of the lumberjack dependency?

File: cmd/start-amazon-cloudwatch-agent/start-amazon-cloudwatch-agent.go [hunk 11]
```diff
@@ -92,8 +93,25 @@ func main() {
...
 	log.Printf("I! Config has been translated into TOML %s \n", tomlConfigPath)
+	echoTOML(tomlConfigPath)
 
 	if err := startAgent(writer); err != nil {
+		log.Printf("E! Error when starting Agent, Error is %v \n", err)
 		os.Exit(1)
 	}
 }
+
+func echoTOML(path string) {
```
Comment (SaxyPandaBear) [inline cmd/start-amazon-cloudwatch-agent/start-amazon-cloudwatch-agent.go:104]: Adam already commented on the naming, which I agree with. I am mildly concerned about the `log.Fatal()` used here. I see above that there is a call directly to `os.Exit()`. Would that be preferable to causing a panic here?

File: go.mod [hunk 12]
```diff
@@ -1,10 +1,11 @@
...
 
-go 1.13
+go 1.17
 
-replace github.com/influxdata/telegraf => github.com/aws/telegraf v0.10.2-0.20201015165757-4470de2d306b
+replace github.com/influxdata/telegraf => github.com/aws/telegraf v0.10.2-0.20220502160831-c20ebe67c5ef
 
-replace github.com/shirou/gopsutil => github.com/aws/telegraf/patches/gopsutil v0.0.0-20201015165757-4470de2d306b // Temporary fix, pending PR https://github.com/shirou/gopsutil/pull/957
+// Temporary fix, pending PR https://github.com/shirou/gopsutil/pull/957
+replace github.com/shirou/gopsutil/v3 => github.com/aws/telegraf/patches/gopsutil/v3 v3.0.0-20220502160831-c20ebe67c5ef // indirect
```
Comment (SaxyPandaBear) [inline go.mod:8]: Is this going to be addressed or should we just expect this to stay here indefinitely

File: internal/k8sCommon/k8sclient/endpoint_test.go [hunk 13]
```diff
@@ -4,13 +4,13 @@
 package k8sclient
 
 import (
+	"gotest.tools/v3/assert"
```
Comment (SaxyPandaBear) [inline internal/k8sCommon/k8sclient/endpoint_test.go:7]: Should probably change this to `"github.com/stretchr/testify/assert"`

File: internal/k8sCommon/k8sclient/pod.go [hunk 14]
```diff
@@ -41,9 +42,11 @@ func (c *podClient) NamespaceToRunningPodNum() map[string]int {
...
 		c.Init()
 	}
-	if c.store.Refreshed() {
-		c.refresh()
-	}
+	//if c.store.Refreshed() {
+	//	log.Printf("I! store refresh %v", c.store.refreshed)
+	//	c.refresh()
+	//}
+	c.refresh()
```
Comment (SaxyPandaBear) [inline internal/k8sCommon/k8sclient/pod.go:49]: Is this a breaking code change in the k8s code?

File: internal/k8sCommon/k8sclient/pod_test.go [hunk 15]
```diff
@@ -4,11 +4,11 @@
 package k8sclient
 
 import (
+	"gotest.tools/v3/assert"
```
Comment (SaxyPandaBear) [inline internal/k8sCommon/k8sclient/pod_test.go:7]: Should be `"github.com/stretchr/testify/assert"`

File: internal/k8sCommon/k8sclient/replicaset_test.go [hunk 16]
```diff
@@ -4,9 +4,9 @@
 package k8sclient
 
 import (
+	"gotest.tools/v3/assert"
```
Comment (SaxyPandaBear) [inline internal/k8sCommon/k8sclient/replicaset_test.go:7]: Should be `"github.com/stretchr/testify/assert"`

File: plugins/inputs/cadvisor/mesos/plugin.go [hunk 17]
```diff
@@ -0,0 +1,41 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
```
Comment (SaxyPandaBear) [inline plugins/inputs/cadvisor/mesos/plugin.go:1]: I'm just going to have to trust this. This is all lifted from another plugin, right?

File: plugins/processors/ec2tagger/ec2tagger.go [hunk 18]
```diff
@@ -316,6 +316,7 @@ func (t *Tagger) Init() error {
 		}
 	}
 
+	t.Log.Infof("ec2tagger: Check ec2 metadata")
```
Comment (SaxyPandaBear) [inline plugins/processors/ec2tagger/ec2tagger.go:319]: 😄 Sounds like this addresses #463

Comment (khanhntd) [reply plugins/processors/ec2tagger/ec2tagger.go:319]: 👍  Thought I would address it in my PR but seems too late for me.

File: cmd/amazon-cloudwatch-agent/register_event_logger_notwindows.go (hunk 10 above)
Comment (ZhenyuTan-amz) [reply cmd/amazon-cloudwatch-agent/register_event_logger_notwindows.go:11]: Will add some explanatory comment for this.

File: go.mod (hunk 12 above)
Comment (ZhenyuTan-amz) [reply go.mod:8]: For now there is no good solution going forward, we will just keep it up-to-date like telegraf, so this will be here going forward.

File: plugins/processors/ec2tagger/ec2tagger.go (hunk 18 above)
Comment (ZhenyuTan-amz) [reply plugins/processors/ec2tagger/ec2tagger.go:319]: I think there is room for improvement with doing #463, this log should be warning level and add more details.

File: plugins/parsers/registry.go (hunk 6 above)
Comment (ZhenyuTan-amz) [reply plugins/parsers/registry.go:29]: Because the method in Parsers 
`func Add(name string, creator Creator) {
`
is taking a Creator as second argument, and the method EMFParserCreator, CSMParserCreator is implementing the Creator interface, https://github.com/aws/telegraf/blob/aws-cwa-release-1-23-0/plugins/parsers/registry.go#L26, so I feel it's more appropriate to be named with xxxParserCreator, while newXXXParser is fine by me as well, I was leaning towards the Creator naming just to reflecting it's implementing that interface.
Let me know if that make sense to you please.

File: cmd/start-amazon-cloudwatch-agent/start-amazon-cloudwatch-agent.go (hunk 5 above)
Comment (ZhenyuTan-amz) [reply cmd/start-amazon-cloudwatch-agent/start-amazon-cloudwatch-agent.go:116]: Debugging log, should be removed.

File: amazon-cloudwatch-container-insights/cloudwatch-agent-dockerfile/source/Dockerfile [hunk 19]
```diff
@@ -6,12 +6,16 @@ FROM $GO_IMAGE as builder
...
 RUN mkdir -p /go/src/github.com/aws/amazon-cloudwatch-agent/
 WORKDIR /go/src/github.com/aws/amazon-cloudwatch-agent/
 
 ARG GO111MODULE="on"
 ENV GO111MODULE=${GO111MODULE}
 
+COPY go.mod /go/src/github.com/aws/amazon-cloudwatch-agent/
+COPY go.sum /go/src/github.com/aws/amazon-cloudwatch-agent/
+RUN go mod download -x
```
Comment (khanhntd) [inline amazon-cloudwatch-container-insights/cloudwatch-agent-dockerfile/source/Dockerfile:18]: Based on my experience, `go mod download` would download more dependencies than actually needed for the build. However, in step [make build-for-docker](https://github.com/aws/amazon-cloudwatch-agent/pull/437/files#diff-45b77cac272377144caa2b7f83dd47cbc6b66943a7f7049879fd6eec4a791ff3R20), there would be a `go build` which build enough dependencies for the build. Therefore, would be better if we delete `go mod download -x`.

File: amazon-cloudwatch-container-insights/k8s-deployment-manifest-templates/deployment-mode/service/cwagent-prometheus/prometheus-eks.yaml [hunk 20]
```diff
@@ -47,8 +50,8 @@ data:
                   ]
                 },
                 {
-                  "source_labels": ["Service", "backend"],
-                  "label_matcher": ".*haproxy-ingress-.*metrics;(httpback-shared-backend|httpback-default-backend|httpsback-shared-backend|_default_backend|.*backend_http)",
+                  "source_labels": ["Service"],
```
Comment (khanhntd) [inline amazon-cloudwatch-container-insights/k8s-deployment-manifest-templates/deployment-mode/service/cwagent-prometheus/prometheus-eks.yaml:53]: Would this be updated in [cloudwatch container insights repository](https://github.com/aws/amazon-cloudwatch-agent/tree/master/amazon-cloudwatch-container-insights) and also in public documents?

File: amazon-cloudwatch-container-insights/k8s-yaml-templates/cwagent-kubernetes-monitoring/otel-config.yml (hunk 9 above)
Comment (khanhntd) [reply amazon-cloudwatch-container-insights/k8s-yaml-templates/cwagent-kubernetes-monitoring/otel-config.yml:1]: Somehow, I would not agree on adding EKS ADOT Config for our repository based on two things:
* Even though ADOT Collector is the side agent, ADOT Collector has a supported EKS on their own. Therefore, it wasteful for us to support this too on EKS.
* In the future in moving to Unified Agent and also inherits ADOT framework, this would be not needed since converting to OTEL Metrics would be under the hood and ADOT Collector would not be used alongside with CWAgent anymore.
Therefore, I would consider removing it instead of adding it.

File: cmd/amazon-cloudwatch-agent/amazon-cloudwatch-agent.go [hunk 21]
```diff
@@ -304,6 +305,7 @@ func runAgent(ctx context.Context,
 		RotationInterval:    ag.Config.Agent.LogfileRotationInterval,
 		RotationMaxSize:     ag.Config.Agent.LogfileRotationMaxSize,
 		RotationMaxArchives: ag.Config.Agent.LogfileRotationMaxArchives,
+		LogWithTimezone:     "",
```
Comment (khanhntd) [inline cmd/amazon-cloudwatch-agent/amazon-cloudwatch-agent.go:308]: What would happen if the LogWithTimeZone is empty string? Would we always want to set UTC Timezone as previous behavior or would we consider moving to Local Timezone to match consistency Customer's Timezone?

File: cmd/amazon-cloudwatch-agent/amazon-cloudwatch-agent.go [hunk 22]
```diff
@@ -47,6 +46,8 @@ import (
 
 const (
 	defaultEnvCfgFileName = "env-config.json"
+	LogTargetEventLog = "eventlog"
```
Comment (khanhntd) [inline cmd/amazon-cloudwatch-agent/amazon-cloudwatch-agent.go:49]: Would consider moving these constants to the `register_event_logger_windows.go` or a separate constant files (same as [SSM Agent](https://github.com/aws/amazon-ssm-agent/blob/mainline/agent/appconfig/constants.go))

File: cmd/start-amazon-cloudwatch-agent/start-amazon-cloudwatch-agent.go [hunk 23]
```diff
@@ -92,8 +93,25 @@ func main() {
 		log.Fatalf("E! Cannot translate JSON config into TOML, ERROR is %v \n", err)
 	}
 	log.Printf("I! Config has been translated into TOML %s \n", tomlConfigPath)
+	echoTOML(tomlConfigPath)
```
Comment (khanhntd) [inline cmd/start-amazon-cloudwatch-agent/start-amazon-cloudwatch-agent.go:96]: My two cents told me we should not log Toml File. What would be the result of printing the Toml File?
1. For customer, JSON Configuration would be enough for them to consider before moving to Toml File 
2. For us, TOML Configuration only helps us debug which is not always the time and we could do it by ourself instead of logging for our customer.

File: internal/k8sCommon/k8sclient/pod.go (hunk 14 above)
Comment (khanhntd) [reply internal/k8sCommon/k8sclient/pod.go:49]: Same question for me 👍

File: plugins/inputs/cadvisor/cadvisor.go [hunk 24]
```diff
@@ -1,8 +1,8 @@
 // Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: MIT
 
-//go:build linux || darwin
-// +build linux darwin
+//go:build linux
```
Comment (khanhntd) [inline plugins/inputs/cadvisor/cadvisor.go:4]: For cadvisor plugin, would we consider adding build constraint for windows?

File: plugins/inputs/cadvisor/mesos/plugin.go (hunk 17 above)
Comment (khanhntd) [reply plugins/inputs/cadvisor/mesos/plugin.go:1]: Same argument with me.

**Review by khanhntd — COMMENTED** [review]
Are there any breaking changes we need to consider besides what we have covered so far regarding the release https://github.com/influxdata/telegraf/releases? Another small question from me would be AWS SDK Go V2 would be addressed in another PR, right? Besides from the previous comments, the release can be roll back and there are no breaking changes that I have seen so far.

File: cmd/amazon-cloudwatch-agent/amazon-cloudwatch-agent.go (hunk 21 above)
Comment (ZhenyuTan-amz) [reply cmd/amazon-cloudwatch-agent/amazon-cloudwatch-agent.go:308]: Using UTC was the existing behavior, using an empty string will result in fall back to UTC, hence maintain the same behavior.

File: cmd/amazon-cloudwatch-agent/amazon-cloudwatch-agent.go (hunk 22 above)
Comment (ZhenyuTan-amz) [reply cmd/amazon-cloudwatch-agent/amazon-cloudwatch-agent.go:49]: Centralize constants is arguably a good idea, but I think organize them into a single constant class should be address in a separated "refactor" PR, here it's simply just following existing pattern for consistence.

File: plugins/inputs/prometheus_scraper/metrics_type_handler.go [hunk 25]
```diff
@@ -40,19 +40,19 @@ func normalizeMetricName(name string, suffixes []string) string {
 }
 
 func (pm *PrometheusMetric) isCounter() bool {
-	return pm.metricType == textparse.MetricTypeCounter
+	return pm.metricType == string(textparse.MetricTypeCounter)
```
[the remark this answers is not in the document]
Comment (ZhenyuTan-amz) [reply plugins/inputs/prometheus_scraper/metrics_type_handler.go:43]: Here is a breaking change from dependency, instead of returning string, it's now returning a enum of **MetricType**, note this is **not** an interface, so I'm casting it to string to keep it work with the PrometheusMetric struct.

File: plugins/inputs/prometheus_scraper/metric_type_handler_test.go [hunk 26]
```diff
@@ -325,24 +325,24 @@ func TestNewMetricsTypeHandler_HandleWithMetricSuffix(t *testing.T) {
 		metricNameBeforeRelabel: "m1_sum",
 		jobBeforeRelabel:        "job1",
 		instanceBeforeRelabel:   "instance1",
-		metricType:              textparse.MetricTypeCounter,
-		tags:                    map[string]string{"job": "job1", "instance": "instance1", "prom_metric_type": textparse.MetricTypeCounter},
+		metricType:              string(textparse.MetricTypeCounter),
```
[the remark this answers is not in the document]
Comment (ZhenyuTan-amz) [reply plugins/inputs/prometheus_scraper/metric_type_handler_test.go:328]: Same thing as my reply in other comment.

File: plugins/inputs/prometheus_scraper/metric_type_handler_test.go [hunk 27]
```diff
@@ -219,16 +219,16 @@ func TestNewMetricsTypeHandler_HandleWithNormalTarget(t *testing.T) {
 		metricNameBeforeRelabel: "m1",
 		jobBeforeRelabel:        "job1",
 		instanceBeforeRelabel:   "instance1",
-		metricType:              textparse.MetricTypeCounter,
-		tags:                    map[string]string{"job": "job1", "instance": "instance1", "prom_metric_type": textparse.MetricTypeCounter},
+		metricType:              string(textparse.MetricTypeCounter),
```
[the remark this answers is not in the document]
Comment (ZhenyuTan-amz) [reply plugins/inputs/prometheus_scraper/metric_type_handler_test.go:222]: same in the other thread.

File: plugins/outputs/cloudwatch/cloudwatch_test.go [hunk 28]
```diff
@@ -394,7 +394,7 @@ func TestWriteError(t *testing.T) {
 
 	tags := map[string]string{"dimension_name1": "dimension_value2"}
 	ti := time.Now()
-	m, _ := metric.New(measurement, tags, fields, ti)
+	m := metric.New(measurement, tags, fields, ti)
```
[the remark this answers is not in the document]
Comment (ZhenyuTan-amz) [reply plugins/outputs/cloudwatch/cloudwatch_test.go:397]: 1. I'm not entirely confident it is not used. Is there business logic within 	`cloudWatchOutput.Write(metrics)` ? My bet will be yes without dig deep into it. And this is our of scope for this PR.

2. Doesn't hurt to use a real time stamp in test IMO, just like using arbitrary strings for dimension names. Not seeing a lot of value to change it.

File: Makefile (hunk 1 above)
Comment (sethAmazon) [reply Makefile:54]: Yeah this needs to be reverted back this was just shortened for testing

File: cmd/start-amazon-cloudwatch-agent/start-amazon-cloudwatch-agent.go (hunk 5 above)
Comment (sethAmazon) [reply cmd/start-amazon-cloudwatch-agent/start-amazon-cloudwatch-agent.go:116]: I like having the toml config output on start up of app but I can remove.

File: amazon-cloudwatch-container-insights/k8s-yaml-templates/cwagent-kubernetes-monitoring/otel-config.yml (hunk 9 above)
Comment (sethAmazon) [reply amazon-cloudwatch-container-insights/k8s-yaml-templates/cwagent-kubernetes-monitoring/otel-config.yml:1]: I can remove this. I was only using this for testing against otel. We are removing otel from cw agent so this should not be here.

File: plugins/inputs/cadvisor/mesos/plugin.go (hunk 17 above)
Comment (sethAmazon) [reply plugins/inputs/cadvisor/mesos/plugin.go:1]: I just copied this directly from cadvisor.

File: plugins/inputs/cadvisor/cadvisor.go (hunk 24 above)
Comment (sethAmazon) [reply plugins/inputs/cadvisor/cadvisor.go:4]: Cadvisor is for linux only. If we want to support windows we would have to use something else.

File: amazon-cloudwatch-container-insights/cloudwatch-agent-dockerfile/source/Dockerfile (hunk 7 above)
Comment (SaxyPandaBear) [reply amazon-cloudwatch-container-insights/cloudwatch-agent-dockerfile/source/Dockerfile:9]: Does this impact anything if for example I decided to build the Docker image on my home desktop?

File: plugins/inputs/cadvisor/mesos/plugin.go (hunk 17 above)
Comment (SaxyPandaBear) [reply plugins/inputs/cadvisor/mesos/plugin.go:1]: There are a lot of instances where we directly take plugin source code and drop it into our repo. Is that a requirement? Is there no way to reference it externally and import it?

File: internal/k8sCommon/k8sclient/pod.go (hunk 14 above)
Comment (sethAmazon) [reply internal/k8sCommon/k8sclient/pod.go:49]: @ZhenyuTan-amz can you test if this still works after removing the comments. I only tested in this state.

Comment (SaxyPandaBear) [reply internal/k8sCommon/k8sclient/pod.go:49]: I don't think removing the comments has any consequence on the execution. I was just asking why it changed because I don't recall seeing a `refresh()` function for the pod client.

**Review by khanhntd — APPROVED** [review]
Not a blocker from me:
* Since this has been worked from internal integration test. However, the `build for linux` has been failed https://github.com/aws/amazon-cloudwatch-agent/runs/6776217563?check_suite_focus=true. 
* Please also run `make fmt` after approval to avoid dilute PR though.
Besides that, LGTM.

File: amazon-cloudwatch-container-insights/k8s-deployment-manifest-templates/deployment-mode/service/cwagent-prometheus/prometheus-eks.yaml (hunk 20 above)
Comment (ZhenyuTan-amz) [reply amazon-cloudwatch-container-insights/k8s-deployment-manifest-templates/deployment-mode/service/cwagent-prometheus/prometheus-eks.yaml:50]: https://github.com/aws-samples/amazon-cloudwatch-container-insights/pull/97/files it's going to be there, need someone to approve it.

File: internal/k8sCommon/k8sclient/pod.go (hunk 14 above)
Comment (ZhenyuTan-amz) [reply internal/k8sCommon/k8sclient/pod.go:49]: Tested and it works with the code commented out. I'm not concerning removing comment would affect anything.

File: internal/k8sCommon/k8sclient/endpoint_test.go (hunk 13 above)
Comment (ZhenyuTan-amz) [reply internal/k8sCommon/k8sclient/endpoint_test.go:7]: "github.com/stretchr/testify/assert" does not support the DeepEqual() used in this test, probably used to support that but not anymore after the version bump.

File: .github/workflows/build-test-linux.yml [hunk 29]
```diff
@@ -41,7 +41,7 @@ jobs:
 
     - uses: zencargo/github-action-go-mod-tidy@v1
       with:
-        go-version: 1.15
+        go-version: 1.17
```
[the remark this answers is not in the document]
Comment (ZhenyuTan-amz) [reply .github/workflows/build-test-linux.yml:44]: Will update

File: amazon-cloudwatch-container-insights/k8s-yaml-templates/cwagent-kubernetes-monitoring/cwagent-configmap.yaml [hunk 30]
```diff
@@ -4,7 +4,6 @@ data:
   # Configuration is in Json format. No matter what configure change you make,
   # please keep the Json blob valid.
   cwagentconfig.json: |
-    {
```
Comment (khanhntd) [inline amazon-cloudwatch-container-insights/k8s-yaml-templates/cwagent-kubernetes-monitoring/cwagent-configmap.yaml:7]: Nitpick -  Missing {

File: logger/lumberjack_const_test.go [hunk 31]
```diff
@@ -102,55 +96,11 @@ func TestWriteToTruncatedFile(t *testing.T) {
 	assert.Equal(t, f[19:], []byte("Z I! SHOULD BE FIRST\n"))
 }
 
-func TestWriteToFileInRotation(t *testing.T) {
```
Comment (SaxyPandaBear) [inline logger/lumberjack_const_test.go:105]: Why remove this test? I'm assuming it's related to a change in lumberjack during the upgrade but what's the consequence?

File: cmd/start-amazon-cloudwatch-agent/start-amazon-cloudwatch-agent.go [hunk 32]
```diff
@@ -92,8 +93,25 @@ func main() {
...
 	if err := startAgent(writer); err != nil {
+		log.Printf("E! Error when starting Agent, Error is %v \n", err)
 		os.Exit(1)
 	}
 }
+
+func printFileContents(path string) {
+	file, err := os.Open(path)
+	if err != nil {
+		log.Fatal(err)
```
Comment (SaxyPandaBear) [inline cmd/start-amazon-cloudwatch-agent/start-amazon-cloudwatch-agent.go:107]: I think my original comment got lost. 

Do we want to `log.Fatal` here and cause a panic, or do we want to gracefully exit with `os.Exit()` like in the above function?

File: logger/lumberjack_const_test.go (hunk 31 above)
Comment (ZhenyuTan-amz) [reply logger/lumberjack_const_test.go:105]: Because with the new change that register lumberjack logger inside Telegraf, we can no longer create the logger like it does in this test 
```
lumberjackLogger := &lumberjack.Logger{
		Filename:   config.Logfile,
		MaxSize:    maxsize,
		MaxBackups: 5,
		MaxAge:     1,
		Compress:   false,
	}
```
Also I restored NewTelegrafWriter() to private as it was in Telegraf, we shouldn't be change it to public and use it, influxdb is not maintaining it as a public method, and we use it anyway which result in it breaks us in this upgrade. So with all those changes, it's hard to test this here properly. I did manually verify the rotation works as it was before, and I start to question the value of this unit test, it's not like any part of the lubmerjack log rotation could be changed by us, so what unit feature is this test covering?

File: internal/k8sCommon/k8sclient/endpoint.go [hunk 33]
```diff
@@ -131,6 +134,7 @@ func (c *epClient) Init() {
 
 	lw := createEndpointListWatch(Get().ClientSet, metav1.NamespaceAll)
 	reflector := cache.NewReflector(lw, &v1.Endpoints{}, c.store, 0)
+	klog.SetLogger(klogr.New().WithName("k8s_client_runtime").V(3))
```
[the remark this answers is not in the document]
Comment (ZhenyuTan-amz) [reply internal/k8sCommon/k8sclient/endpoint.go:137]: I think smaller number means higher log level, 0 is fatal, 2 is info, 3 is Extended information about changes, 5 is trace level and 6 is Display requested resources. I don't know why Seth chose 3 here, but seems appropriate to me.

File: amazon-cloudwatch-container-insights/k8s-yaml-templates/cwagent-kubernetes-monitoring/cwagent-configmap.yaml (hunk 30 above)
[the remark this answers is not in the document]
Comment (ZhenyuTan-amz) [reply amazon-cloudwatch-container-insights/k8s-yaml-templates/cwagent-kubernetes-monitoring/cwagent-configmap.yaml:7]: Will add it back

File: cmd/start-amazon-cloudwatch-agent/start-amazon-cloudwatch-agent.go (hunk 32 above)
Comment (ZhenyuTan-amz) [reply cmd/start-amazon-cloudwatch-agent/start-amazon-cloudwatch-agent.go:107]: Not my code but I think we should keep consistency here, I will change it to follow the existing pattern.

**Review by khanhntd — DISMISSED** [review]
LGTM with all these changes.

File: cmd/start-amazon-cloudwatch-agent/start-amazon-cloudwatch-agent.go [hunk 34]
```diff
@@ -92,8 +93,26 @@ func main() {
...
+
+func printFileContents(path string) {
+	file, err := os.Open(path)
+	if err != nil {
+		log.Printf("E! Error when printing file(%s) contents, Error is %v \n", path, err)
+		os.Exit(1)
+	}
+	defer func() {
+		if err = file.Close(); err != nil {
+			log.Fatal(err)
```
Comment (SaxyPandaBear) [inline cmd/start-amazon-cloudwatch-agent/start-amazon-cloudwatch-agent.go:112]: I don't recall what happens in this scenario. If we `defer` here, what are the repercussions of panicking?

File: cmd/start-amazon-cloudwatch-agent/start-amazon-cloudwatch-agent.go [hunk 35]
```diff
@@ -92,8 +93,26 @@ func main() {
...
+		}
+	}()
+
+	b, err := ioutil.ReadAll(file)
```
Comment (SaxyPandaBear) [inline cmd/start-amazon-cloudwatch-agent/start-amazon-cloudwatch-agent.go:116]: Do we ignore the error? Sorry I missed this the first time.

Comment (ZhenyuTan-amz) [reply cmd/start-amazon-cloudwatch-agent/start-amazon-cloudwatch-agent.go:116]: Logged an error for this.

**Review by SaxyPandaBear — APPROVED** [review]
LGTM
